File: bindings/node/examples/quickstart.js
/**
 * CybouDB Node.js & TypeScript Quickstart Demo
 *
 * Demonstrates:
 * - Unified single-file .cdb database
 * - Relational tables and B+Tree indexes
 * - Durable FIFO queues
 * - Replayable event streams
 * - Unified cross-primitive ACID transactions
 */

const fs = require('fs');
const path = require('path');
const { Database, version } = require('../index.js');

const DB_PATH = path.join(__dirname, 'quickstart_demo.cdb');

function main() {
  if (fs.existsSync(DB_PATH)) {
    fs.unlinkSync(DB_PATH);
  }

  console.log('=== CybouDB Node.js Quickstart Demo ===');
  console.log(`-> Engine Version: ${version()}`);

  // 1. Create database (512 pages = 2 MiB)
  console.log(`-> Creating database: ${DB_PATH}`);
  const db = Database.create(DB_PATH, 512);

  try {
    // 2. Define schema
    console.log('-> Creating schema (Table, Index, Queue, Stream, Cursor)...');
    db.execute('CREATE TABLE workers (id INT32 NOT NULL, name TEXT, tasks INT32);');
    db.execute('CREATE UNIQUE INDEX idx_worker_id ON workers (id);');
    db.execute('CREATE QUEUE tasks;');
    db.execute('CREATE STREAM audit_log;');
    db.execute('CREATE CURSOR monitor ON audit_log;');

    // 3. Populate workers
    console.log('-> Populating workers with parameterized query...');
    db.execute('INSERT INTO workers VALUES (?, ?, ?);', [1, 'Worker Alpha', 0]);
    db.execute('INSERT INTO workers VALUES (?, ?, ?);', [2, 'Worker Beta', 0]);

    // 4. Enqueue jobs
    console.log("-> Enqueueing jobs into 'tasks' FIFO queue...");
    db.enqueue('tasks', 'generate_embeddings');
    db.enqueue('tasks', 'index_documents');

    // 5. Atomic transaction across Queue + Table + Stream
    console.log('-> Executing atomic transaction across Queue + Table + Stream...');
    db.transaction((tx) => {
      // Dequeue task
      const taskBytes = tx.dequeue('tasks');
      const taskName = taskBytes ? taskBytes.toString('utf-8') : 'unknown';
      console.log(`   [Tx] Dequeued job: '${taskName}'`);

      // Update worker record
      tx.execute('UPDATE workers SET tasks = 1 WHERE id = 1;');
      console.log('   [Tx] Incremented completed tasks for Worker #1');

      // Append audit event
      tx.append('audit_log', `Worker Alpha finished '${taskName}'`);
      console.log('   [Tx] Appended audit event to stream');

      console.log('   [Tx] Committing transaction atomically...');
    });

    // 6. Query updated state
    console.log('-> Querying workers from table:');
    const rows = db.query('SELECT id, name, tasks FROM workers WHERE id = ?;', [1]);
    for (const row of rows) {
      console.log(`   Worker #${row[0]}: Name='${row[1]}', Tasks=${row[2]}`);
    }

    // 7. Read stream events
    console.log("-> Reading audit event from stream as cursor 'monitor':");
    const event = db.readStream('audit_log', 'monitor');
    if (event) {
      console.log(`   Stream Event: ${event.toString('utf-8')}`);
    }
  } finally {
    db.close();
    if (fs.existsSync(DB_PATH)) {
      fs.unlinkSync(DB_PATH);
    }
  }

  console.log('=== Demo completed successfully! ===');
}

main();

File: bindings/python/examples/quickstart.py
"""
CybouDB Python Quickstart Example

Demonstrates:
- Creating a unified .cdb database file
- Relational tables and secondary B+tree indexes
- Durable FIFO queues
- Replayable event streams
- Unified cross-primitive ACID transactions
"""

import os
import cyboudb

DB_PATH = "quickstart_demo.cdb"

def main():
    # Clean up old demo file if present
    if os.path.exists(DB_PATH):
        os.remove(DB_PATH)

    print("=== CybouDB Python Quickstart Demo ===")
    print(f"-> Engine Version: {cyboudb.version()}")

    # 1. Create a database (512 pages = 2 MiB)
    print(f"-> Creating database: {DB_PATH}")
    db = cyboudb.Database.create(DB_PATH, pages=512)

    try:
        # 2. Schema definition
        print("-> Creating schema (Table, Index, Queue, Stream, Cursor)...")
        db.execute("CREATE TABLE workers (id INT32 NOT NULL, name TEXT, completed INT32);")
        db.execute("CREATE UNIQUE INDEX idx_workers_id ON workers (id);")
        db.execute("CREATE QUEUE tasks;")
        db.execute("CREATE STREAM audit_log;")
        db.execute("CREATE CURSOR monitor ON audit_log;")

        # 3. Populate data
        print("-> Populating workers with parameterized statements...")
        db.execute("INSERT INTO workers VALUES (?, ?, ?);", [1, "Worker Alpha", 0])
        db.execute("INSERT INTO workers VALUES (?, ?, ?);", [2, "Worker Beta", 0])

        # 4. Enqueue jobs
        print("-> Enqueueing tasks into 'tasks' FIFO queue...")
        db.enqueue("tasks", "sync_embeddings")
        db.enqueue("tasks", "generate_summary")

        # 5. Atomic cross-primitive transaction
        print("-> Executing atomic transaction across Queue + Table + Stream...")
        with db.transaction() as tx:
            # Dequeue next task
            task_bytes = tx.dequeue("tasks")
            task_str = task_bytes.decode("utf-8") if task_bytes else "unknown"
            print(f"   [Tx] Dequeued job: '{task_str}'")

            # Update relational worker record
            tx.execute("UPDATE workers SET completed = 1 WHERE id = 1;")
            print("   [Tx] Updated completed count for Worker #1")

            # Append audit event to stream
            tx.append("audit_log", f"Worker Alpha completed '{task_str}'")
            print("   [Tx] Appended audit log event to stream")

            # Block commits automatically at exit; rolls back on exception
            print("   [Tx] Committing transaction atomically...")

        # 6. Query updated state
        print("-> Querying workers from table:")
        rows = db.query("SELECT id, name, completed FROM workers WHERE id = ?;", [1])
        for row in rows:
            print(f"   Worker #{row[0]}: Name='{row[1]}', Completed={row[2]}")

        # 7. Read stream events
        print("-> Reading audit event from stream as cursor 'monitor':")
        event = db.read_stream("audit_log", "monitor")
        if event:
            print(f"   Stream Event: {event.decode('utf-8')}")

    finally:
        db.close()
        if os.path.exists(DB_PATH):
            os.remove(DB_PATH)

    print("=== Demo completed successfully! ===")

if __name__ == "__main__":
    main()
